File: 병사배치하기.py
'''
7
15 11 4 8 5 2 4

뒤집어서 LIS 로 푼다
7
4 2 5 8 4 11 15
'''

n = int(input())
arr = list(map(int, input().split()))
arr.reverse()

dp = [1] * n # DP 테이블을 1로 초기화

# 가장 긴 증가하는 부분 수열(LIS: Longest Increasing Sequence) 알고리즘 적용
for i in range(1,n):  # 현재 체크하려는 원소
    for j in range(0, i):  # 현재 원소와 비교하여 체크하여야 하는 이전 원소
        if arr[i] > arr[j]:
            dp[i] = max(dp[i], dp[j] + 1)

print(n-max(dp))

File: 병사배치하기.java
/**
 내림차순 전투력이 되게끔 하는 최소 열외 병사 수를 구하는 문제
 
 7
 15 11 4 8 5 2 4
 */

import java.util.*;
public class 병사배치하기 {
    static int n;
    static ArrayList<Integer> v = new ArrayList<Integer>();
    static int[] dp = new int[2000];
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        for (int i = 0; i < n; i++) {
            v.add(sc.nextInt());
        }
        // 배열의 순서를 뒤집어 LIS로 푼다.
        Collections.reverse(v);
        for (int i = 0; i < n; i++) dp[i] = 1;
        for (int i = 1; i < n; i++) 
            for(int j = 0; j < i; j++)
                if(v.get(i) > v.get(j)) dp[i] = Math.max(dp[i], dp[j] + 1);
        int max = 0;
        for(int i = 0; i < n; i++) max = Math.max(max, dp[i]);
        System.out.println(n-max);

    }
    
}